### PR TITLE
EndpointChannels allow pre-computing stuff upfront

### PR DESCRIPTION
**Pull request by iamdanfox**
## Before this PR

We're leaving performance on the table by recomputing things based on endpoint info (e.g. in the UserAgent channel, we do some string concatenation on every single request based on the Endpoint serviceName). This pattern happens in a bunch of places (e.g. building up tagged metrics using metric schema). Given that we know the complete set of endpoints at codegen time, we should be able to do something like Witchcraft's approach of binding everything upfront.

## After this PR: 30% speedup
==COMMIT_MSG==
Codegen can take advantage of a new methods to do endpoint-specific work once upfront, instead of spending CPU cycles doing it again and again on every request.
==COMMIT_MSG==

This does introduce some new public API:

- `EndpointChannel`
- `ChannelEndpointStage`
- some methods on the `runtime.clients()` object

After merging this, I'd need to go and change the conjure-java codegen to use this.

```
Benchmark                             Mode  Cnt      Score      Error  Units
EndToEndBenchmark.dialogue2Blocking  thrpt   16  25647.619 ± 3542.047  ops/s
EndToEndBenchmark.dialogueBlocking   thrpt   16  19812.490 ± 3081.670  ops/s
```

## Possible downsides?
- From an API-design perspective, we now have an alternative way of doing the same thing as `Channel`, which could be confusing from a maintenance perspective
- We have one `instanceof` check which sends people down a different codepath if they pass in a bog standard `Channel` vs if they pass in a `ChannelEndpointStage`
- Until I can land the corresponding codegen in conjure-java, this will cause every RPC call to allocate a O(tens) more objects, so I think there will be a slight perf hit.

### CODE REVIEW

File: dialogue-example/dialogue-example-dialogue/src/main/java/com/palantir/dialogue/example/SampleServiceAsync2.java
Comment (iamdanfox) [inline dialogue-example/dialogue-example-dialogue/src/main/java/com/palantir/dialogue/example/SampleServiceAsync2.java:88]: open question - should we actually just bind the `voidToVoidDeserializer` upfront too? This would save us 1 (one) allocation of `deserializer::deserialize`

File: dialogue-target/src/main/java/com/palantir/dialogue/BindEndpoint.java
Comment (iamdanfox) [inline dialogue-target/src/main/java/com/palantir/dialogue/BindEndpoint.java]: this is public because `DefaultClients` lives in `:dialogue-serde` and doesn't depend on `:dialogue-core`, but needs to be able to bind DialogueClients.

Comment (carterkozak) [reply dialogue-target/src/main/java/com/palantir/dialogue/BindEndpoint.java]: Could we write things through the runtime object? I'd like the runtime to be capable of taking a channel and endpoint, and returning an EndpointChannel.

File: dialogue-example/dialogue-example-dialogue/src/main/java/com/palantir/dialogue/example/SampleServiceAsync2.java
Comment (iamdanfox) [inline dialogue-example/dialogue-example-dialogue/src/main/java/com/palantir/dialogue/example/SampleServiceAsync2.java]: @carterkozak I think the PR does the thing you describe? here we're saying to the runtime here's a `Channel` and and `Endpoint`, please give me an EndpointChannel.

File: dialogue-target/src/main/java/com/palantir/dialogue/BindEndpoint.java
Comment (iamdanfox) [reply dialogue-target/src/main/java/com/palantir/dialogue/BindEndpoint.java]: Is this kinda what you were imagining? https://github.com/palantir/dialogue/pull/764/files#r426760705

Comment (iamdanfox) [reply dialogue-target/src/main/java/com/palantir/dialogue/BindEndpoint.java]: We could move this particular interface if we're happy to change some inter-project dependencies (e.g. if we made `:core` depend on `:serde` or vice versa)

File: dialogue-example/dialogue-example-dialogue/src/main/java/com/palantir/dialogue/example/SampleServiceAsync2.java
Comment (carterkozak) [reply dialogue-example/dialogue-example-dialogue/src/main/java/com/palantir/dialogue/example/SampleServiceAsync2.java]: Got it. I'm a bit worried about the instanceof checks, I think it may be safer to take a semantic break in the runtime and require the bind method instead of directly executing against the channel.

File: .palantir/revapi.yml
Comment (carterkozak) [inline .palantir/revapi.yml]: we also document `no guarantees are made for custom implementations` :-)

File: dialogue-core/src/main/java/com/palantir/dialogue/core/DeprecationWarningChannel.java
Comment (carterkozak) [inline dialogue-core/src/main/java/com/palantir/dialogue/core/DeprecationWarningChannel.java]: super minor nit: may want to use something concrete with a toString for debugging.

File: dialogue-example/dialogue-example-dialogue/src/main/java/com/palantir/dialogue/example/SampleServiceAsync2.java
Comment (iamdanfox) [reply dialogue-example/dialogue-example-dialogue/src/main/java/com/palantir/dialogue/example/SampleServiceAsync2.java]: Ok I think I've got rid of all the `instanceof` checks apart from the one in the runtime (which seems pretty unavoidable to me), as we still want people to be able to pass in a bog-standard `Channel` and have their FooBlocking still work

```java
        if (channel instanceof ChannelEndpointStage) {
            return ((ChannelEndpointStage) channel).endpoint(endpoint);
        }
```

File: dialogue-core/src/main/java/com/palantir/dialogue/core/DeprecationWarningChannel.java
Comment (iamdanfox) [reply dialogue-core/src/main/java/com/palantir/dialogue/core/DeprecationWarningChannel.java]: Yep I'm very much aware that I added all these nice toStrings and now lambdas & anonymous classes will have ruined them all.

File: dialogue-target/src/main/java/com/palantir/dialogue/EndpointChannel.java
Comment (iamdanfox) [inline dialogue-target/src/main/java/com/palantir/dialogue/EndpointChannel.java]: maybe the class used to construct one of these should actually be called an `EndpointChannelBuilder`?

File: dialogue-target/src/main/java/com/palantir/dialogue/Clients.java
Comment (iamdanfox) [inline dialogue-target/src/main/java/com/palantir/dialogue/Clients.java:38]: Thoughts on just deprecating this guy?

File: dialogue-target/src/main/java/com/palantir/dialogue/EndpointChannel.java
Comment (ferozco) [reply dialogue-target/src/main/java/com/palantir/dialogue/EndpointChannel.java]: Perhaps `EndpointChannelFactory`?

File: dialogue-core/src/main/java/com/palantir/dialogue/core/NeverThrowChannel.java
Comment (ferozco) [inline dialogue-core/src/main/java/com/palantir/dialogue/core/NeverThrowChannel.java:61]: I'm not a huge fan of all of this duplication :(

File: dialogue-core/src/main/java/com/palantir/dialogue/core/DialogueChannel.java
Comment (ferozco) [inline dialogue-core/src/main/java/com/palantir/dialogue/core/DialogueChannel.java]: nit: rename to something more descriptive, perhaps: `CompositeChannel`?

**Review by ferozco — COMMENTED** [review]
just some quick thoughts, going to play around with it for a bit

File: dialogue-core/src/main/java/com/palantir/dialogue/core/NeverThrowChannel.java
Comment (iamdanfox) [reply dialogue-core/src/main/java/com/palantir/dialogue/core/NeverThrowChannel.java:61]: Yeah this is because this class is used in both the per-uri 'channel' style (below the NSS) and also in the 'precompiled endpoint' style above the NSS.

I _think_ it might actually make sense to move the NeverThrow logic into the runtime, because actually if users provide a non-sensical Channel implementation it would actually be nice to still catch mistakes and turn them into failedFutures...

**Comment by ferozco** [comment]
quick sketch of a different approach that consolidates the channel factory logic into a single location instead of being within each channel: https://github.com/palantir/dialogue/compare/dfox/endpoint-channel...fo/proposal?expand=1

File: dialogue-serde/src/main/java/com/palantir/conjure/java/dialogue/serde/DefaultClients.java
Comment (iamdanfox) [inline dialogue-serde/src/main/java/com/palantir/conjure/java/dialogue/serde/DefaultClients.java]: type name changed. let's switch to debug

File: dialogue-target/src/main/java/com/palantir/dialogue/EndpointChannel.java
Comment (carterkozak) [inline dialogue-target/src/main/java/com/palantir/dialogue/EndpointChannel.java]: nit: wouldn't necessarily note how this is constructed

File: dialogue-core/src/main/java/com/palantir/dialogue/core/ActiveRequestInstrumentationChannel.java
Comment (ferozco) [inline dialogue-core/src/main/java/com/palantir/dialogue/core/ActiveRequestInstrumentationChannel.java:60]: why do we need an EndpointChannel and Channel version of this? Should they just be separate classes rather than nested?

File: dialogue-example/dialogue-example-dialogue/src/main/java/com/palantir/dialogue/example/SampleServiceAsync2.java
Comment (ferozco) [inline dialogue-example/dialogue-example-dialogue/src/main/java/com/palantir/dialogue/example/SampleServiceAsync2.java:55]: did we decide to go with a separate factory method in codegen?

File: dialogue-target/src/main/java/com/palantir/dialogue/EndpointChannelFactory.java
Comment (ferozco) [inline dialogue-target/src/main/java/com/palantir/dialogue/EndpointChannelFactory.java:20]: nit: update docs

File: dialogue-core/src/main/java/com/palantir/dialogue/core/DialogueChannel.java
Comment (ferozco) [inline dialogue-core/src/main/java/com/palantir/dialogue/core/DialogueChannel.java]: nit: rename to `CompositeChannel`

File: dialogue-core/src/test/java/com/palantir/dialogue/core/ContentDecodingChannelTest.java
Comment (ferozco) [inline dialogue-core/src/test/java/com/palantir/dialogue/core/ContentDecodingChannelTest.java:47]: it is nice to reduce a bit of boilerplate in tests :)

File: dialogue-core/src/test/java/com/palantir/dialogue/core/RetryingChannelTest.java
Comment (ferozco) [inline dialogue-core/src/test/java/com/palantir/dialogue/core/RetryingChannelTest.java:62]: assume the test changes are just refactors, so i'm just skimming

File: dialogue-serde/src/main/java/com/palantir/conjure/java/dialogue/serde/DefaultClients.java
Comment (ferozco) [inline dialogue-serde/src/main/java/com/palantir/conjure/java/dialogue/serde/DefaultClients.java:86]: could avoid the extra allocation by inlining the dispatch in `call`

File: dialogue-serde/src/main/java/com/palantir/conjure/java/dialogue/serde/DefaultClients.java
Comment (ferozco) [inline dialogue-serde/src/main/java/com/palantir/conjure/java/dialogue/serde/DefaultClients.java:195]: duplicate class?

File: dialogue-example/dialogue-example-dialogue/src/main/java/com/palantir/dialogue/example/SampleServiceBlocking2.java
Comment (iamdanfox) [inline dialogue-example/dialogue-example-dialogue/src/main/java/com/palantir/dialogue/example/SampleServiceBlocking2.java:29]: Corresponding codegen change: https://github.com/palantir/conjure-java/pull/910

File: dialogue-core/src/main/java/com/palantir/dialogue/core/ActiveRequestInstrumentationChannel.java
Comment (carterkozak) [reply dialogue-core/src/main/java/com/palantir/dialogue/core/ActiveRequestInstrumentationChannel.java:60]: We use this for two separate stages, once around wire requests, and once around top-level 'processing' requests that includes those waiting in a queue/retry.

Comment (iamdanfox) [reply dialogue-core/src/main/java/com/palantir/dialogue/core/ActiveRequestInstrumentationChannel.java:60]: I think separately I'd like to revaluate all our metrics and figure out the cost/benefit for each of them (both in $ and CPU)

File: dialogue-core/src/test/java/com/palantir/dialogue/core/RetryingChannelTest.java
Comment (iamdanfox) [reply dialogue-core/src/test/java/com/palantir/dialogue/core/RetryingChannelTest.java:62]: yup, didn't add/remove any of em.

File: dialogue-serde/src/main/java/com/palantir/conjure/java/dialogue/serde/DefaultClients.java
Comment (iamdanfox) [reply dialogue-serde/src/main/java/com/palantir/conjure/java/dialogue/serde/DefaultClients.java:195]: Not quite, one is in :serde and the other is in :core, we don't have a dep in either direction. I named em the same just because they're identical

**Comment by ferozco** [comment]
Looks reasonable to me, I'll let @carterkozak take a look too before we ship this

File: dialogue-target/src/main/java/com/palantir/dialogue/Clients.java
Comment (carterkozak) [inline dialogue-target/src/main/java/com/palantir/dialogue/Clients.java:37]: This may cause issues compiling generated code before the generator is upgraded. May want to hold off a day or two before adding a deprecation.

File: dialogue-target/src/main/java/com/palantir/dialogue/EndpointChannelFactory.java
Comment (carterkozak) [inline dialogue-target/src/main/java/com/palantir/dialogue/EndpointChannelFactory.java:28]: Not sure how undefined this can really be -- I think we'll need to do something along those lines for the feign shim, but we're not terribly concerned with feign performance as everything moves to pure dialogue.

**Review by carterkozak — APPROVED** [review]

**Pull request merged**
